src/bindings/python/flux/job/watcher.py

```diff
@@ -0,0 +1,574 @@
+#############################################################
+# Copyright 2023 Lawrence Livermore National Security, LLC
+# (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+#
+# This file is part of the Flux resource manager framework.
+# For details, see https://github.com/flux-framework.
+#
+# SPDX-License-Identifier: LGPL-3.0
+##############################################################
+
+import os
+import sys
+import time
+from collections import Counter
+
+import flux
+from flux.job import output_watch_async
+from flux.progress import ProgressBar
+
+
+class JobStatus:
+    """
+    Simple convenience class for caching job "state" in the JobProgressBar
+    and JobWatch classes.
+
+    Attributes:
+        id: The job id
+        status: This job's current sinmplified status (See below for possible
+            status values)
+        exitcode: The exit code of the job (0 for success, otherwise failure)
+
+    Current valid statuses include:
+    - pending
+    - running
+    - complete
+    - failed
+    """
+
+    def __init__(self, job):
+        """
+        Initialize a JobStatus object using an instance of flux.job.JobInfo
+        """
+        if not isinstance(job, flux.job.JobInfo):
+            raise ValueError("JobStatus requires an object of type JobInfo")
+        self.id = job.id
+        self.status, self.exitcode = self._jobinfo_get_status_and_exitcode(job)
+        self._events = Counter()
+
+    def _jobinfo_get_status_and_exitcode(self, job):
+        #  Return simplified job status and exitcode if job is inactive
+        status = job.status
+        if status in ("DEPEND", "PRIORITY", "SCHED"):
+            return "pending", 0
+        elif status in ("RUN", "CLEANUP"):
+            return "running", 0
+        elif status == "COMPLETED":
+            return "complete", 0
+        elif status in ("FAILED", "CANCELED", "TIMEOUT"):
+            rc = job.returncode
+            if not rc:
+                rc = 1
```

Comment:
An empty returncode can occur for a job that failed before a `start` event due to a non cancel type exception.
However, your comment made me realize the fix should really be applied to the  `JobInfo.returncode` property function, since as you said, that result would be surprising. I'll push a commit that does that and remove this little block.